tutorial/practicals/nemo/4_nemo_openacc/README.md

```diff
@@ -0,0 +1,416 @@
+parallelise any loop nests within the marked-up region. In PSyclone
+this is achieved by applying the [`ACCKernelsTrans`][ref_kernelstrans]
+transformation to suitable regions of the code. The advantage of this
+approach is that it minimises the number of directives that must be
+inserted and makes use of the compiler's own dependency analysis to
+ensure that loops may be safely parallelised. This means that
+for the tracer advection mini-app we could simply enclose each
+top-level loop (that does not contain a CodeBlock) within a KERNELS
+region:
+
+```python
+    for node in sched.children:
+        if isinstance(node, Loop):
+	    try:
+	        ACC_KERNELS_TRANS.apply(node)
+	    except TransformationError:
+	        pass
+```
+
+In doing this we will enclose the outer, 'iteration' loop within a
+KERNELS region. If we look at the original Fortran (in
+`tra_adv_mod.F90`) we can see that this loop cannot be parallelised
+because it both reads and writes the `mydomain` array so that each
+iteration depends upon the results of the previous one. We would
+therefore be relying upon the OpenACC compiler to "do the right thing"
+and parallelise the loops *within* the iteration loop.
+
+However, the point of a Domain-Specific Language is that it makes use
+of domain-specific knowledge and therefore we can be more prescriptive
+about how we want the compiler to parallelise the code (in order to
+achieve better computational performance). There are also things that
+cannot be put inside KERNELS regions: these include calls to other
+subroutines and certain types of statement that are known to trigger
+compiler bugs (e.g. `MIN(my_array(:,:,:), dim=2)` causes problems with
+NVIDIA (PGI) < 20.7).
+
+It is therefore necessary to exercise fine-grained control when
+inserting KERNELS regions and this tutorial will walk through some of
+the steps. (If you already have OpenACC experience you will see that
+some of these steps are not going to result in performant code,
+however, they are included here to help students grasp the various
+ways in which PSyclone can be used.)
+
+### 1. Enclose loops over vertical levels ###
+
+1. Modify the supplied `kernels_trans.py` optimisation script to apply the
+   `ACCKernelsTrans` transformation to every loop over vertical levels
+   within the outer, iteration loop of the mini-app. The script already
+   locates that loop:
+
+   ```python
+    # Find the outer, 'iteration' loop
+    tloop = None
+    for node in sched.children:
+        if isinstance(node, Loop) and node.loop_type == "tracers":
+            tloop = node
+            break
+    ```
+
+   Similar to what has been done in previous tutorials, you will need to
+   loop over the children of that loop, identify those that are loops of
+   the correct "levels" type, and transform them:
+   ```python
+    for child in tloop.loop_body.children:
+        if isinstance(node, Loop) and node.loop_type == "levels":
```

Comment:
both references to "node" here should be "child"

Reply:
Oops! Changed 'child' to 'node' as that was quicker.